src/typecheck/mod.rs

```diff
@@ -1326,121 +1659,189 @@ pub fn unify_(
             )),
         },
         (TypeWrapper::Ptr(p1), TypeWrapper::Ptr(p2)) if p1 == p2 => Ok(()),
         // The two following cases are not merged just to correctly distinguish between the
         // expected type (first component of the tuple) and the inferred type when reporting a row
         // unification error.
         (TypeWrapper::Ptr(p), tyw) => {
-            constr_unify(state.constr, p, &tyw)
-                .map_err(|err| err.into_unif_err(TypeWrapper::Ptr(p), tyw.clone()))?;
-            state.table.insert(p, Some(tyw));
+            constr_unify(state.constr, p, &tyw, &env2).map_err(|err| {
+                err.into_unif_err(
+                    Closure::new(TypeWrapper::Ptr(p), env1),
+                    Closure::new(tyw.clone(), env2.clone()),
+                )
+            })?;
+            state.table.insert(p, Some(Closure { ty: tyw, env: env2 }));
             Ok(())
         }
         (tyw, TypeWrapper::Ptr(p)) => {
-            constr_unify(state.constr, p, &tyw)
-                .map_err(|err| err.into_unif_err(tyw.clone(), TypeWrapper::Ptr(p)))?;
-            state.table.insert(p, Some(tyw));
+            constr_unify(state.constr, p, &tyw, &env1).map_err(|err| {
+                err.into_unif_err(
+                    Closure::new(tyw.clone(), env1.clone()),
+                    Closure::new(TypeWrapper::Ptr(p), env2),
+                )
+            })?;
+            state.table.insert(p, Some(Closure { ty: tyw, env: env1 }));
             Ok(())
         }
         (TypeWrapper::Constant(i1), TypeWrapper::Constant(i2)) if i1 == i2 => Ok(()),
         (TypeWrapper::Constant(i1), TypeWrapper::Constant(i2)) => {
             Err(UnifError::ConstMismatch(i1, i2))
         }
-        (ty, TypeWrapper::Constant(i)) | (TypeWrapper::Constant(i), ty) => {
-            Err(UnifError::WithConst(i, ty))
-        }
+        (ty, TypeWrapper::Constant(i)) => Err(UnifError::WithConst(i, Closure::new(ty, env1))),
+        (TypeWrapper::Constant(i), ty) => Err(UnifError::WithConst(i, Closure::new(ty, env2))),
     }
 }
 
 /// Try to unify two row types. Return an [`IllformedRow`](./enum.RowUnifError.html#variant.IllformedRow) error if one of the given type
 /// is not a row type.
 pub fn unify_rows(
     state: &mut State,
-    t1: AbsType<Box<TypeWrapper>>,
-    t2: AbsType<Box<TypeWrapper>>,
+    t1c: Closure, //AbsType<Box<TypeWrapper>>,
+    t2c: Closure, //AbsType<Box<TypeWrapper>>,
+                  //env: TypeEnvironment
 ) -> Result<(), RowUnifError> {
+    let Closure { ty: t1w, env: env1 } = t1c;
+    let Closure { ty: t2w, env: env2 } = t2c;
+    let t1 = if let TypeWrapper::Concrete(absty) = t1w {
+        absty
+    } else {
+        unreachable!("Not concreate abstype unify_rows (1)")
+    };
+    let t2 = if let TypeWrapper::Concrete(absty) = t2w {
+        absty
+    } else {
+        unreachable!("Not concreat abstype unify_rows (2)")
+    };
```

Comment:
Rather than having this, you can just pass the original abs type and the environment separated. Not all arguments need to be passed as packed closures. E.g:

```
unify_rows(
  state: &mut State,
  t1: AbsType<Box<TypeWrapper>>,
  env1: Environment,
  t2: AbsType...
```

Then you don't have those nasty `unreachable` anymore.